File: 4 - Replace Conditional Logic with Strategy/C#/MPG.ReplaceConditionalLogicWithStrategy/MPG.ReplaceConditionalLogicWithStrategy.After/CapitalStrategy.cs
using System;
using System.Linq;

namespace MPG.ReplaceConditionalLogicWithStrategy.After
{
    public abstract class CapitalStrategy
    {
        private const int MillisPerDay = 86400000;
        private const int DaysPerYear = 365;

        public abstract double Capital(Loan loan);

        public virtual double Duration(Loan loan)
        {
            return YearsTo(loan.GetExpiry().Value, loan);
        }

        protected static double RiskFactor(Loan loan)
        {
            return RiskFactors.ForRating(loan.GetRiskRating());
        }

        protected static double UnusedRiskFactor(Loan loan)
        {
            return UnusedRiskFactors.ForRating(loan.GetRiskRating());
        }

        private double WeightedAverageDuration(Loan loan)
        {
            var duration = 0.0;
            var weightedAverage = loan.GetPayments().Sum(payment => YearsTo(payment.Date, loan) * payment.Amount);
            var sumOfPayments = loan.GetPayments().Sum(payment => payment.Amount);

            if (loan.GetCommitment() != 0.0)
            {
                duration = weightedAverage / sumOfPayments;
            }

            return duration;
        }

        protected double YearsTo(DateTime endDate, Loan loan)
        {
            var beginDate = loan.GetToday().HasValue ? loan.GetToday().Value : loan.GetStart().Value;
            var beginDateMilliseconds = new DateTimeOffset(beginDate).ToUnixTimeMilliseconds();
            var endDateMilliseconds = new DateTimeOffset(endDate).ToUnixTimeMilliseconds();
            return ((endDateMilliseconds - beginDateMilliseconds) / MillisPerDay) / DaysPerYear;
        }
    }
}

File: 5 - Form Template Method/Python/before/capital_strategy.py
from abc import abstractmethod
from risk_factors import RiskFactors
from unused_risk_factors import UnusedRiskFactors
from dateutil.relativedelta import relativedelta


class CapitalStrategy:
    @abstractmethod
    def capital(self, loan):
        raise NotImplementedError

    def duration(self, loan):
        return self.years_to(loan.get_expiry(), loan)

    @staticmethod
    def risk_factor_for(loan):
        return RiskFactors.for_rating(loan.get_risk_rating())

    @staticmethod
    def unused_risk_factor_for(loan):
        return UnusedRiskFactors.for_rating(loan.get_risk_rating())

    @staticmethod
    def years_to(end_date, loan):
        begin_date = loan.get_today() if loan.get_today() is not None else loan.get_start()
        difference = relativedelta(end_date, begin_date)
        return difference.years
